### ISSUE TITLE
insert_image with stream decodes the image to get basic image properties, maybe allow providing them

### ISSUE THREAD

**Issue by MerlijnWajer** [pymupdf/PyMuPDF#1037]
**Is your feature request related to a problem? Please describe.**

Since PyMuPDF 1.18.8, insert_image seems to be a bit slower. It creates a pixmap of the stream to get the width, height and alpha channel. It looks like it previously used a faster method (I might be wrong). It is my understanding that mupdf will *also* decode the image (by creating an image instance) when inserting the image, so I think PyMuPDF essentially decodes the image twice. This is not such a big problem with JPG or PNG images, but with JPEG2000 images, decoding can take a while, so this hurts insert performance.

Locally I "solved" this problem with a hack:

```

diff --git a/fitz/utils.py b/fitz/utils.py
index e40019e..c60fc9e 100644
--- a/fitz/utils.py
+++ b/fitz/utils.py
@@ -384,12 +384,18 @@ def insertImage(*args, **kwargs) -> None:
         alpha = pixmap.alpha
         digest = calc_hash(pixmap.samples)
     elif stream:
-        pix = Pixmap(stream)
-        w = pix.width
-        h = pix.height
-        alpha = pix.alpha
-        digest = calc_hash(pix.samples)
-        del pix
+        if 'ww' in kwargs:
+            w = kwargs.get('ww')
+            h = kwargs.get('hh')
+            alpha = kwargs.get('alphaalpha')
+            digest = calc_hash(stream)
+        else:
+            pix = Pixmap(stream)
+            w = pix.width
+            h = pix.height
+            alpha = pix.alpha
+            digest = calc_hash(pix.samples)
+            del pix
         if type(stream) is io.BytesIO:
             stream = stream.getvalue()
     elif xref > 0:
```

**Describe the solution you'd like**

Perhaps these properties can be provided to PyMuPDF's insert_image call so that the image doesn't need to be decoded. I realise it's not pretty, so there might be alternative ways to do this.

**Additional context**

Migrating from PyMuPDF 1.18.2 to 1.18.13 caused degraded insert_image performance for JPEG2000 images, and I believe this is the cause.

**Comment by JorjMcKie** [comment]
You are right, I had to change something to get rid of a pesky memory leak in that version. This action caused the other problem you recently reported.
All that goes back to the fact that I am generating the PDF matrix for displaying the image in Python instead of in C. So I need image properties width / height / alpha already in Python.
If I would undergo the effort to compute that matrix in C, the need to open the image twice would go away.
Frankly, I would prefer going this way. Let me check this out.

**Comment by JorjMcKie** [comment]
@MerlijnWajer - I am still evaluating the best way to address this very valid request.
My current thinking. Refacturing the code in C should be the way to go, but it does not make "expert" parameter options obsolete. In fact, the more is known about an image, the more ways are possible to speed up the insertion because obviously 
Currently, if `stream` and `mask` are both provided, I do not check for alpha anywhere, because the situation already is obvious.
Being given width and height information, that would also be beneficial in the same sense.

So, in short: yes I will add width, height and alpha as additional parameters.

Thanks for bringing this up!

**Comment by JorjMcKie** [comment]
Ok, done it.
Now, every image is only accessed once for insertion. Width / height, corresponding matrix and MD5 digest are computed in C code only.
Specification of width, height and alpha may be specified. Width/height will have no performance improvement effect.
If an image is known to have no alpha channel, then specifying ``alpha=0`` will significantly improve response time, because the corresponding check is quite costly.
If specifying a separate ``mask``, then this will overrule transparency checking as it was before.

Probably won't concern you, but there also is a new `xref` parameter specifying an existing image anywhere in the file. If present, then no image access at all will take place, and necessary information will be taken from that object.
If xref=0, the image will be MD5-hashed (for which it must be read) to look up any previous copy in the file. If found, a reference only will be inserted.

I am currently testing the above. All looks good so far, and performance improvements are significant!

**Comment by JorjMcKie** [comment]
Fixed in v1.18.13, which was just published.
